tools/purge_ems.rb

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env ruby
+require File.expand_path('../config/environment', __dir__)
+require 'optimist'
+
+ARGV.shift if ARGV[0] == '--' # if invoked with rails runner
+batch_message = "For each relationship, the number of rows to be removed in a batch. A lower number may decrease memory usage but also take longer."
+opts = Optimist.options do
+  banner "Purge management system records in the background via the queue.\n\nUsage: ruby #{$PROGRAM_NAME} [options]\n\nOptions:\n\t"
+  opt :id,      "Management System id",                                         :type => :integer,                    :required => true
+  opt :follow,  "Follow the progress or exit after creating work items.",       :type => :boolean, :default => true
+  opt :batch,   batch_message,                                                  :type => :integer, :default => 1_000
+  opt :timeout, "The number of minutes each work item will be allowed to run.", :type => :integer, :default => 30
+end
+
+Optimist.die :id,      "must be a positive number"                             if opts[:id] < 1
+Optimist.die :batch,   "must be a positive number"                             if opts[:batch] < 1
+Optimist.die :timeout, "must be a positive number greater than or equal to 10" if opts[:timeout] < 10
+
+DESTROY_METHOD = "destroy".freeze
+ZONE           = nil                    # Will queue for any zone
+PRIORITY       = MiqQueue::LOW_PRIORITY # Don't block other work
+
+def log(msg)
+  $log.info("MIQ(#{__FILE__}) #{msg}")
+  puts msg
+end
+
+def current_backlog
+  MiqQueue.where(:method_name => DESTROY_METHOD, :zone => ZONE, :priority => PRIORITY).count
+end
+
+def before_destroy_ems(ems_id)
+  ems = ExtManagementSystem.find(ems_id)
+  if ems.enabled?
+    log("Pausing management system with id: #{ems.id} #{ems.name}...")
+    ems.pause!
+    log("Pausing management system with id: #{ems.id} #{ems.name}...Complete")
+  end
+
+  log("Removing any workers for management system with id: #{ems.id} #{ems.name}...")
+  ems.ems_workers.each(&:kill_async)
+  ems.wait_for_ems_workers_removal
+  log("Removing any workers for management system with id: #{ems.id} #{ems.name}...Complete")
+end
+
+def existing_backlog?
+  # Watching the destroy queue messages be processed can take a long time so terminal disconnects
+  # are likely.  Therefore, if running it again, detect when it's in progress, skip queueing more
+  # and skip right to watching the updated progress if desired.
+  current_backlog > 0
+end
+
+def destroy_ems_in_batches(ems_id, batch, timeout, priority)
+  before_destroy_ems(ems_id)
+
+  if existing_backlog?
+    log("A backlog of work items is already in progress, skipping creating more.")
+    return
+  end
+
+  ems = ExtManagementSystem.find(ems_id)
```

Comment:
Yeah, I wasn't sure it made sense to capture stale ems association data as we delete other emses.  If it's a performance hit, maybe but I would prefer to do them one at a time just in case. I'm not sure.

yes, we can have a more helpful message if the ems didn't exist.  I considered printing the ems id and name of all emses in that case but wasn't sure if that was a good idea.